### PR TITLE
feat: add timepb support library

### PR DESCRIPTION
**Pull request by robert-zaremba**
## Summary 

Moved from: https://github.com/cosmos/cosmos-sdk/pull/11085

A support library to do time operation with [protobuf Timestamp](https://github.com/protocolbuffers/protobuf-go/tree/master/types/known/timestamppb)

## Motivation

With new proto directions we are going away from go stdlib `time.Time` to the protobuf Timestamppb package.
We often do time operations (eg comparing times) and we need a canonical library for Timestamppb  rather then repeating / copying code. 

## Updates
* @fdymylja suggested to put it in `types/wellknown/time`. We are not creating a new types, but a support library, so I propose here different path: `support/timepb`
* Originally I used `pbtime`, but it seams in protobuf they prefer to add `pb` as a suffix.

### CODE REVIEW

**Comment by robert-zaremba** [comment]
Let's decide about `IsZero`. If we don't want to have a special meaning for a zero time (not a nil), then I prefer to remove it.

**Review by aaronc — COMMENTED** [review]
Generally looks good, I'm still not sure about the zero check but I see the docs make clear what "zero" means

File: support/timepb/cmp.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+		return 0
+	}
+	if t1.Seconds < t2.Seconds || t1.Seconds == t2.Seconds && t1.Nanos < t2.Nanos {
+		return -1
+	}
+	return 1
+}
+
+func DurationIsNegative(d durpb.Duration) bool {
+	return d.Seconds < 0 || d.Seconds == 0 && d.Nanos < 0
```
Comment (technicallyty) [inline support/timepb/cmp.go:32]: not sure how durations are calculated between seconds + nanos, but would the following invalidate this?:

```go
	d := dur.Duration{
		Seconds: 1,
		Nanos:   -100000000000000000000000,
	}
```

also my IDE is complaining that the parameter should be a pointer

Comment (robert-zaremba) [reply support/timepb/cmp.go:32]: This is invalid duration, it won't pass `CheckValid()`.
I don't think we should care with invalid values here, they should be checked upfront.

Comment (robert-zaremba) [reply support/timepb/cmp.go:32]: > also my IDE is complaining that the parameter should be a pointer

I'm curious why? `durpb.New` indeed returns a pointer, but he structure is small so there is no need to use pointers. But I will update and use pointer in this function.

**Comment by robert-zaremba** [comment]
@aaronc  so maybe we should remove the `IsZero`?  (ref: https://github.com/cosmos/cosmos-proto/pull/60#issuecomment-1028751992)

File: support/timepb/cmp.go (hunk 1 above)
Comment (robert-zaremba) [reply support/timepb/cmp.go:32]: On the other hand it feels weird to use `time.Duration` without pointer and `durpb.Durration` with pointer.

**Comment by robert-zaremba** [comment]
I've added rapid tests

File: support/timepb/cmp.go (hunk 1 above)
Comment (robert-zaremba) [reply support/timepb/cmp.go:34]: nice.
I've already updated to use pointers.

File: support/timepb/cmp.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+// DurationIsNegative returns true if the duration is negative. It assumes that d is valid
+// (d..CheckValid() is nil).
+func DurationIsNegative(d *durpb.Duration) bool {
+	return d.Seconds < 0 || d.Seconds == 0 && d.Nanos < 0
+}
+
+// AddStd returns a new timestamp with value t + d, where d is stdlib Duration.
+// If t is nil then nil is returned.
+// Panics on overflow.
+func AddStd(t *tspb.Timestamp, d time.Duration) *tspb.Timestamp {
```
Comment (technicallyty) [inline support/timepb/cmp.go:40]: I think i would prefer to have this called `AddDuration` or `AddStdDuration`,

**Review by technicallyty — APPROVED** [review]
looks good, just 1 nit

File: support/timepb/cmp.go (hunk 2 above)
Comment (robert-zaremba) [reply support/timepb/cmp.go:40]: In stdlib we have `time.Add` not `time.AddDuration`. You can't add time to another time, it doesn't make sense.

Comment (aaronc) [reply support/timepb/cmp.go:40]: I think the short naming is okay

**Comment by aaronc** [comment]
> @aaronc so maybe we should remove the `IsZero`? (ref: [#60 (comment)](https://github.com/cosmos/cosmos-proto/pull/60#issuecomment-1028751992))

I'd say let's drop `IsZero`. Other stuff looks good

File: support/timepb/README.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# timepb
```
Comment (aaronc) [inline support/timepb/README.md:1]: Should this be a README or docs.go? Not sure what is best practice. I think the advantage of docs.go is that it shows up in pkg.go.dev like this for sub-packages:
image

Comment (robert-zaremba) [reply support/timepb/README.md:1]: README shows up in pkg.go.dev as well when you go to the package page. I think we need both. Will add doc.go

File: support/timepb/doc.go [hunk 4]
```diff
@@ -0,0 +1,5 @@
+/*
+Package timepb provides functions to do time operations with protobuf timestamp
+and duration  structures.
```
Comment (aaronc) [inline support/timepb/doc.go:3]:
```suggestion
and duration structures.
```

**Review by aaronc — APPROVED** [review]

**Review by technicallyty — APPROVED** [review]

**Pull request merged**